cdap-authorization/src/main/java/co/cask/common/authorization/client/DefaultAuthorizationClient.java

```diff
@@ -0,0 +1,180 @@
+import co.cask.common.authorization.SubjectId;
+import co.cask.common.authorization.UnauthorizedException;
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicate;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import com.google.inject.Inject;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Default implementation of {@link AuthorizationClient} which uses {@link ACLStore}.
+ */
+public class DefaultAuthorizationClient implements AuthorizationClient {
+
+  private final Supplier<ACLStore> aclStoreSupplier;
+
+  /**
+   * @param aclStoreSupplier used to set and get {@link co.cask.common.authorization.ACLEntry}s
+   */
+  @Inject
+  public DefaultAuthorizationClient(ACLStoreSupplier aclStoreSupplier) {
+    this.aclStoreSupplier = aclStoreSupplier;
+  }
+
+  public DefaultAuthorizationClient(ACLStore aclStore) {
+    this.aclStoreSupplier = Suppliers.ofInstance(aclStore);
+  }
+
+  @Override
+  public void authorize(Iterable<ObjectId> objects, Iterable<SubjectId> subjects,
+                        Iterable<Permission> requiredPermissions) throws UnauthorizedException {
+    if (!isAuthorized(objects, subjects, requiredPermissions)) {
+      throw new UnauthorizedException(objects, subjects, requiredPermissions);
+    }
+  }
+
+  @Override
+  public void authorize(ObjectId object, Iterable<SubjectId> subjects,
+                        Iterable<Permission> requiredPermissions) throws UnauthorizedException {
+    if (!isAuthorized(object, subjects, requiredPermissions)) {
+      throw new UnauthorizedException(Collections.singleton(object), subjects, requiredPermissions);
+    }
+  }
+
+  @Override
+  public boolean isAuthorized(Iterable<ObjectId> objects, Iterable<SubjectId> subjects,
+                              Iterable<Permission> requiredPermissions) {
+
+    // TODO: check isAuthorized of parent objects
+    Set<Permission> remainingRequiredPermission = Sets.newHashSet(requiredPermissions);
+    Set<ObjectId> visitedObjectIds = Sets.newHashSet();
+
+    // TODO: go from most general (GLOBAL -> NAMESPACE -> APP)
+    for (ObjectId object : objects) {
+      // TODO: consider doing only a single aclStore call
+      for (ObjectId parentObject : object.getParents()) {
+        try {
+          checkAuthorized(remainingRequiredPermission, visitedObjectIds, parentObject, subjects);
+          if (remainingRequiredPermission.isEmpty()) {
+            // Early exit if all permissions are fulfilled
+            return true;
+          }
+        } catch (Exception e) {
+          // Ignore since we assume unauthorized if there's an issue getting the ACLs
```

Comment:
This was ignored so that if later calls to `checkAuthorized` satisfy all the permissions, then the user can still be authorized. But, I suppose if HBase is down then there's no point in trying. Just wasn't sure if there could be some other ways to fail that would fail 100% given the same state.